File: easy/112_path_sum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    bool hasPathSum(TreeNode* root, int sum) {
        if (!root) { return false; }
        bool flg = false;
        findPath(root, sum, flg);
        return flg;
    }
    
    void findPath(TreeNode* node, int sum, bool &flg) {
        if (!node->left && !node->right && sum == node-> val) {
            flg = true;
        } 
        if (node->left) { findPath(node->left, sum - node->val, flg); }
        if (node->right) { findPath(node->right, sum - node->val, flg); }
        
    }
};

File: easy/112_path_sum.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    // DFS
    public boolean hasPathSum(TreeNode root, int sum) {
        if (root == null) return false;
        return helper(root, 0, sum);
    }
    
    private boolean helper(TreeNode node, int curSum, int target) {
        if (node == null) { return false; }
        int sum = curSum + node.val;
        // Only return when the node have no child
        if (node.left == null && node.right == null) {
            return sum == target ? true : false;
        }
        return helper(node.left, sum, target) || helper(node.right, sum, target);
    }
}